release-2.1: assorted missing telemetry

knz opened the pull request:
Backport:
  * 4/4 commits from "sql/parser: add missing telemetry for unimplemented features" (#31635)
  * 1/1 commits from "pgwire: add telemetry for fetch limits" (#31637)
  * 1/1 commits from "sql,telemetry: report built-in function name in eval failures" (#31677)


Please see individual PRs for details.

/cc @cockroachdb/release

knz committed:
sql/parser: ensure that CREATE/DROP SCHEMA attempts land in stats

Release note (sql change): Attempts to use `CREATE/DROP SCHEMA` will
now be collected as telemetry if statistics reporting is enabled, to
gauge demand for user-defined schemas.

knz committed:
sql/parser: mark constraint deferrability options for telemetry

Release note (sql change): Attempts to use `DEFERRABLE` and other
constraint deferrability options will now be collected as telemetry if
statistics reporting is enabled, to gauge demand for tuning time of
constraint checking.

knz committed:
sql/parser: track CREATE TABLE (LIKE ...) for telemetry

Release note (sql change): Attempts to use `CREATE TABLE (LIKE...)`
will now be collected as telemetry if statistics reporting is enabled,
to gauge demand for this feature.

knz committed:
sql/parser: track usage of CREATE TABLE ... WITH in telemetry

Release note (sql change): Attempts to use `CREATE TABLE ... WITH`
will now be collected as telemetry if statistics reporting is enabled,
to gauge demand for these features.

knz committed:
pgwire: add telemetry for fetch limits

The JDBC driver and perhaps others commonly try to use the "fetch
limit" parameter, which is yet unsupported in
CockroachDB (#4035). This patch adds telemetry to gauge demand.

Release note (sql change): attempts by client apps to use the
unsupported "fetch limit" parameter (e.g. via JDBC) will now be
captured in telemetry if statistics reporting is enabled, to gauge
support for this feature.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/31638)
<!-- Reviewable:end -->

knz committed:
sql,telemetry: report built-in function name in eval failures

Prior to this patch, if any built-in function was producing a
non-pgerror error object it would be reported in telemetry as
`othererror.eval.go:3573`.

This patch fixes this by ensuring the function name is included and
the error reported as a pg "data error",
e.g. `othererror.22000.somefunction()`.

In addition, this patch extends the connExecutor error handling code
to always include a `pgerror.Error`'s `InternalCommand` in telemetry
if set, not just when the code is "internal error" or "syntax error".

Release note (sql change): CockroachDB will now include the name of
the SQL built-in function in collected statistics upon evaluation
errors.

bdarnell commented on pkg/sql/parser/sql.y:
```diff
@@ -8355,6 +8376,7 @@ unreserved_keyword:
 | DAY
 | DEALLOCATE
 | DELETE
+| DEFERRED
```
Grammar changes including new keywords after RC1? I think this missed the cutoff and needs to wait for 2.2. Telemetry work really needs to be planned as a part of the main development cycle; it's not something we can do post-beta when we're only supposed to be making stability-improving changes. @awoods187

knz replied:
This is an unreserved keyword. It doesn't change the set of queries that will be accepted nor the set of queries that produce errors.

awoods187 commented:
@bdarnell it wasn't planned this late in the cycle it was a minor oversight. It's really important for us to have good telemetry in the initial major release because many users don't upgrade to point releases--or at least not as frequently as we urge them to do so.

knz commented:
@BramGruneir can you review this PR today? This may be good to get into 2.1.1.

BramGruneir reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

As long as this has @bdarnell's blessing.

> Reviewed 2 of 2 files at r1, 3 of 3 files at r2, 2 of 2 files at r3, 3 of 3 files at r4, 4 of 4 files at r5, 4 of 4 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31638#-:-LR8JhiCAT-eUnvdniPA:bm64uu4)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

awoods187 commented:
let's definitely get this into 2.1.1 cc @BramGruneir

knz commented:
TC is wedged on a stress test somehow...

knz commented:
ah, it is done.
